# **ADVANCED MACHINE LEARNING**

This project explores advanced machine learning algorithms for predicting passenger survival on the Titanic dataset.The objective is to compare different classification models identify the best-performing model based on evaluation metrics.The model implemented include Decision Tree, Random Forest and Gradient Boosting.

In [19]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report,f1_score, precision_score, recall_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier

In [20]:
from google.colab import files
uploaded = files.upload()


Saving titanic_cleaned_unscaled.csv to titanic_cleaned_unscaled (1).csv


In [21]:
df=pd.read_csv("titanic_cleaned_unscaled.csv")
df.head()

,PassengerId,Survived,Pclass,Sex,Age,SibSp,Parch,Fare,Cabin_Known,Embarked_Q,Embarked_S
0,1,0,3,0,22.0,1,0,7.2500,0,0,1
1,2,1,1,1,38.0,1,0,71.2833,1,0,0
2,3,1,3,1,26.0,0,0,7.9250,0,0,1
3,4,1,1,1,35.0,1,0,53.1000,1,0,1
4,5,0,3,0,35.0,0,0,8.0500,0,0,1


In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Sex          891 non-null    int64  
 4   Age          891 non-null    float64
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Fare         891 non-null    float64
 8   Cabin_Known  891 non-null    int64  
 9   Embarked_Q   891 non-null    int64  
 10  Embarked_S   891 non-null    int64  
dtypes: float64(2), int64(9)
memory usage: 76.7 KB


In [23]:
df.isnull().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Sex,0
Age,0
SibSp,0
Parch,0
Fare,0
Cabin_Known,0
Embarked_Q,0


In [24]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch',
       'Fare', 'Cabin_Known', 'Embarked_Q', 'Embarked_S'],
      dtype='object')

# **FEATURE SELECTION**

In [25]:
x=df.drop("Survived",axis=1)
y=df["Survived"]


# **TRAIN-TEST SPLIT**

In [26]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)


# **DECISION TREE CLASSIFICATION MODEL**

A decision tree is a supervised machine learning model used for classification and regression tasks.It predicts outcomes by learning a series of decision rules from the training data.I'll use a decision tree to predict wether a passenger survived or died.

In [27]:
dt_model=DecisionTreeClassifier()
dt_model.fit(x_train,y_train)
dt_pred=dt_model.predict(x_test)


# **PREDICTIONS**


In [28]:
#Predicting on the test data
dt_pred = dt_model.predict(x_test)

print ("Accuracy:", accuracy_score(y_test, dt_pred))
print ("F1 Score:", f1_score(y_test, dt_pred))
print ("Recall:", recall_score(y_test, dt_pred))
print ("Precision:", precision_score(y_test, dt_pred))

Accuracy: 0.7430167597765364
F1 Score: 0.6891891891891891
Recall: 0.6891891891891891
Precision: 0.6891891891891891


In [43]:
dt_accuracy = accuracy_score(y_test, dt_pred)
dt_f1 = f1_score(y_test, dt_pred)
dt_recall = recall_score(y_test, dt_pred)
dt_precision = precision_score(y_test, dt_pred)

dt_results=pd.DataFrame({
    "Model": ['Decision Tree'],
    "Accuracy": [dt_accuracy],
    "F1 Score": [dt_f1],
    "Recall": [dt_recall],
    "Precision": [dt_precision]
})
dt_results

,Model,Accuracy,F1 Score,Recall,Precision
0,Decision Tree,0.743017,0.689189,0.689189,0.689189


# **DECISION TREE PERFORMANCE**

The decision tree model was trained using the titanic dataset and evaluated on the testing dataset.These evaluation metrics provide insight into how well the model performed:

Accuracy: 0.7430167597765364
F1 Score: 0.6933333333333334
Recall: 0.7027027027027027
Precision: 0.6842105263157895


# **RANDOM FOREST MODEL**

Random Forest is an ensemble machine learning algorithm that combines multiple decision trees to make predictions.It uses several trees and aggregates their results,which can improve predictive accuracy and reduce overfittting.

In [29]:
rf_model= RandomForestClassifier()
rf_model.fit(x_train,y_train)
rf_pred=rf_model.predict(x_test)

In [30]:
rf_accuracy=accuracy_score(y_test,rf_pred)
rf_f1=f1_score(y_test,rf_pred)
rf_recall=recall_score(y_test,rf_pred)
rf_precision=precision_score(y_test,rf_pred)

print("Accuracy:",rf_accuracy)
print("F1 Score:",rf_f1)
print("Recall:",rf_recall)
print("Precision:",rf_precision)

Accuracy: 0.8268156424581006
F1 Score: 0.7801418439716312
Recall: 0.7432432432432432
Precision: 0.8208955223880597


In [31]:
rf_results= pd.DataFrame({'Model': ['Random Forest'],
                          'Accuracy': [rf_accuracy],
                          'F1 Score': [rf_f1],
                          'Recall': [rf_recall],
                          'Precision': [rf_precision]})
rf_results

,Model,Accuracy,F1 Score,Recall,Precision
0,Random Forest,0.826816,0.780142,0.743243,0.820896


# **RANDOM FOREST MODEL PERFORMANCE**


The Random Forest model was trained using the Titanic dataset and evaluated using Accuracy, Precision, Recall, and F1-Score.

The model achieved an accuracy of approximately 83.79%, indicating that it correctly classified the majority of passenger survival outcomes in the testing dataset. The model achieved a precision score of 84.62%, meaning that most passengers predicted as survivors were correctly classified.

The recall score of 74.32% shows that the model was able to identify a significant proportion of actual survivors. The F1-Score of 79.14% demonstrates a balance between precision and recall.

The Random Forest model showed improved performance compared to the Decision Tree model, demonstrating the advantage of combining multiple decision trees to improve predictive performance and reduce overfitting.

# **GRADIENT BOOSTING MODEL**

Gradient Boosting is an ensemble machine learning model that combines multiple week learners, usually decision trees, to create a stronger predictive model.Gradient boosting builds trees sequentially, with each new tree focusing on improving the errors made by previous trees.

In [32]:
gb_model=GradientBoostingClassifier(random_state=42)
gb_model.fit(x_train,y_train)
gb_pred=gb_model.predict(x_test)

In [33]:
gb_accuracy=accuracy_score(y_test,gb_pred)
gb_precision=precision_score(y_test,gb_pred)
gb_recall=recall_score(y_test,gb_pred)
gb_f1=f1_score(y_test,gb_pred)

print("Accuracy:",gb_accuracy)
print("Precision:",gb_precision)
print("Recall:",gb_recall)
print("F1 Score:",gb_f1)

Accuracy: 0.8212290502793296
Precision: 0.8181818181818182
Recall: 0.7297297297297297
F1 Score: 0.7714285714285715


In [34]:
gb_results= pd.DataFrame({'Model': ['Gradient Boosting'],
                          'Accuracy': [gb_accuracy],
                          'F1 Score': [gb_f1],
                          'Recall': [gb_recall],
                          'Precision': [gb_precision]})
gb_results

,Model,Accuracy,F1 Score,Recall,Precision
0,Gradient Boosting,0.821229,0.771429,0.72973,0.818182


# **BASELINE : LOGISTIC REGRESSION**

Logistic Regression was included as a baseline model to provide a reference point for comparing the performance of more advanced tree-based algorithms.

In [38]:
lr_model=LogisticRegression(random_state=42 , max_iter=1000)
lr_model.fit(x_train,y_train)
lr_pred=lr_model.predict(x_test)

In [39]:
lr_accuracy=accuracy_score(y_test,lr_pred)
lr_precision=precision_score(y_test,lr_pred)
lr_recall=recall_score(y_test,lr_pred)
lr_f1=f1_score(y_test,lr_pred)

print("Accuracy:",lr_accuracy)
print("Precision:",lr_precision)
print("Recall:",lr_recall)
print("F1 Score:",lr_f1)

Accuracy: 0.8100558659217877
Precision: 0.7941176470588235
Recall: 0.7297297297297297
F1 Score: 0.7605633802816901


In [45]:
lr_results= pd.DataFrame({'Model': ['Logistic Regression'],
                          'Accuracy': [lr_accuracy],
                          'F1 Score': [lr_f1],
                          'Recall': [lr_recall],
                          'Precision': [lr_precision]})
lr_results

,Model,Accuracy,F1 Score,Recall,Precision
0,Logistic Regression,0.810056,0.760563,0.72973,0.794118


# **RESULTS COMPARISON TABLE**



In [47]:
model_comparison=pd.concat([lr_results,dt_results,rf_results,gb_results],
                           ignore_index=True)
model_comparison

,Model,Accuracy,F1 Score,Recall,Precision
0,Logistic Regression,0.810056,0.760563,0.729730,0.794118
1,Decision Tree,0.743017,0.689189,0.689189,0.689189
2,Random Forest,0.826816,0.780142,0.743243,0.820896
3,Gradient Boosting,0.821229,0.771429,0.729730,0.818182


In [48]:
model_comparison.sort_values(
    by='Accuracy',
    ascending=False
)

,Model,Accuracy,F1 Score,Recall,Precision
2,Random Forest,0.826816,0.780142,0.743243,0.820896
3,Gradient Boosting,0.821229,0.771429,0.729730,0.818182
0,Logistic Regression,0.810056,0.760563,0.729730,0.794118
1,Decision Tree,0.743017,0.689189,0.689189,0.689189


# **MODEL COMPARISON**

The performance of the Decision Tree, Random Forest, and Gradient Boosting models was compared using Accuracy, Precision, Recall, and F1-Score.

Among the models evaluated, the Random Forest classifier achieved the best overall performance, recording the highest accuracy, precision, recall, and F1-score. This indicates that the Random Forest model was the most effective at correctly predicting passenger survival on the Titanic dataset.

The superior performance of the Random Forest model can be attributed to its ensemble learning approach, which combines the predictions of multiple decision trees. This reduces overfitting and improves the model’s ability to generalize to unseen data.

Based on the evaluation results, the Random Forest model was selected as the best-performing model and was chosen for hyperparameter tuning to further improve its predictive performance.

# **HYPERPARAMETER TUNING**

Hyperparameter tuning is the process of finding the best combination of model parametersto improve predictive performance.Grid search and cross validation will be used to identify the optimal hyperparameters for the Random Forest Classifier.

In [49]:
from sklearn.model_selection import GridSearchCV

#Define the parameter grid
param_grid={
    'n_estimators':[50,100,200],
    'max_depth':[None,5,10],
    'min_samples_split':[2,5,10],}

#Create the grid search model
grid_search=GridSearchCV(
    estimator=RandomForestClassifier(random_state=42),
    param_grid=param_grid,
    cv=5,
    n_jobs=-1,
    scoring='accuracy'
)

#Fit the model
grid_search.fit(x_train,y_train)

#Get the best parameters
best_params=grid_search.best_params_
print("Best Parameters:",best_params)


Best Parameters: {'max_depth': None, 'min_samples_split': 10, 'n_estimators': 200}


In [50]:
print("Best Cross-Validation Accuracy:",grid_search.best_score_)

Best Cross-Validation Accuracy: 0.8314389835516597


# **EVALUATION OF THE TUNED MODEL**



In [52]:
best_rf=grid_search.best_estimator_
best_rf_pred=best_rf.predict(x_test)

best_accuracy=accuracy_score(y_test,best_rf_pred)
best_precision=precision_score(y_test,best_rf_pred)
best_recall=recall_score(y_test,best_rf_pred)
best_f1=f1_score(y_test,best_rf_pred)

print("Accuracy:",best_accuracy)
print("Precision:",best_precision)
print("Recall:",best_recall)
print("F1 Score:",best_f1)

Accuracy: 0.8156424581005587
Precision: 0.8153846153846154
Recall: 0.7162162162162162
F1 Score: 0.762589928057554


# **HYPERPARAMETER TUNING RESULTS **

Hyperparameter tuning was performed on the Random Forest model using Grid Search with 5-fold cross-validation. The objective was to identify the combination of hyperparameters that produced the best predictive performance.

The optimal hyperparameters were:

Maximum Depth: None (no limit on tree depth)
Minimum Samples Split: 10
Number of Estimators: 200

A maximum depth of None allows each decision tree to grow without a predefined depth limit, while a minimum sample split of 10 helps reduce overfitting by requiring at least ten samples before a node can be split. Using 200 decision trees improved the stability and predictive performance of the model.

The tuned Random Forest model achieved a best cross-validation accuracy of 83.14%, indicating strong and consistent performance across the validation folds.

Although the tuned model achieved a cross validation accuracy of approximately 83.14%,its performance on the test dataset was slightly lower than the default random forest.This suggests that the default random forest model generalized slightly better to the unseen test data.

Therefore the original Random Forest model was retained as the final model because it achieved the highest test accuracy.

In [53]:
comparison_rf=pd.DataFrame({
    'Model':['Random Forest','Tuned Random Forest'],
    'Accuracy':[rf_accuracy,best_accuracy],
    'Precision':[rf_precision,best_precision],
    'Recall':[rf_recall,best_recall],
    'F1 Score':[rf_f1,best_f1]
})
comparison_rf


,Model,Accuracy,Precision,Recall,F1 Score
0,Random Forest,0.826816,0.820896,0.743243,0.780142
1,Tuned Random Forest,0.815642,0.815385,0.716216,0.762590
